split up ecsql reference into indiviual files, remove iframed iModel … (#6377)
iTwin/itwinjs-core: docs/learning/ECSqlReference/ECSqlFunctions.md

Following is list of built-in scalar functions

ECSQL allows use of these built-in functions:

1. [`ec_classname()`](#ec_classname-ecclassid--format-string--format-id) - Gets the formatted/qualified class name, given ECClassId as input
1. [`ec_classid())`](#ec_classidschema-name-or-alias----classname) - Gets ECClassId, given a formatted/qualified class name as input
1. [`regexp()`](#regexp--regex-value) - test if a text matches a regex.
1. [`regexp_extract()`](#regexp_extract--value-regex--rewrite) - extract and rewrite matching regex group from a string value.
1. [`strToGuid()`](#strtoguid-guid-string) - covert string guid to binary guid.
1. [`guidToStr()`](#guidtostring-binary-guid) - covert binary guid to string guid.
1. [`navigation_value()`](#navigation_value-ecnavigationproperty-path-id--RelECClassId) - Constructs an ECNavigation value, given ECNavigation property, Id and optionaly, RelECClassId.

## ec_classname( _ecclassId_ [, _format-string_ | _format-id_] )

For the specified ecClassId, returns the class name as a string formatted according to the specified format-string

### Parameters

`ecclassId`: An integer which could be a constant, column or a parameter.

> `format-string | format-id`: Optional format specifier and could be one of the following values. `NULL` is also valid value -- this is the same as not specifying the second parameter at all

Reviewer: `this is the same as not specifying the second parameter at all`

As if the second parameter were not specified? Not sure what the intent is...

### Example

```sql
-- returns schema-name:classname
SELECT ec_classname([ECClassId], 's:c') FROM [BisCore].[Element]

-- same as 'sa:cn' - returns schema-alias:classname
SELECT ec_classname([ECClassId], 1) FROM bis:Element

-- returns schema-name, after filtering on it
SELECT * FROM [BisCore].[Element] WHERE ec_classname([ECClassId], 's') = 'BisCore'

-- returns schema-alias after filtering on it
SELECT * FROM [BisCore].[Element] WHERE ec_classname([ECClassId], 3) = 'bis'

-- only get classname and filter on classname
SELECT * FROM [BisCore].[Element] WHERE ec_classname([ECClassId], 'c') = 'PUMP'
```

## ec_classId('_schema-name-or-alias_ : | . _classname_' )

For the specified (qualified) class name, returns the `ECCassId`.

Note that this function can also take in two arguments - in the following form where _schema-name-or-alias_ and _classname_ can be specified separately.
`ec_classid[ '<schema-name-or-alias>',  '<classname>')`

## Parameters

Can take either one or two parameters:
`schema-name-or-alias`: Schema name or alias e.g. bis (alias) or BisCore (name)
`class-name`: Name of the class e.g. Element

### Returns

> The function return a integer `ECClassId` or `NULL` if the name could not be resolved.

Reviewer: ```suggestion
An integer `ECClassId` or `NULL` if the name could not be resolved.
```

DisplayLabel
--------------------
Terrain Boundary
Terrain Breakline
Terrain Drape Boundary
Terrain Drape Void
Terrain Hole
Terrain Island
Terrain Reference
Terrain Source Contour
Terrain Spot Elevation
Terrain Void
```

## REGEXP_EXTRACT ( _value_, _regex_ [, _rewrite_] )

Regex uses [google/re2](https://github.com/google/re2/wiki/Syntax) engine.
This function can be used to extract or rewrite the output. Parameter `rewrite` is made of group reference where `\0` refer to text captured by whole regex specified. `\1`, `\2` `...` refer to regex capture group in that order.

```sql
-- In follow we rewrite the string by swapping first and second capture group
SELECT
    REGEXP_EXTRACT(DisplayLabel,'(\w+)\s+(\w+)', '\2,\1')
FROM meta.ECClassDef c
    WHERE REGEXP('Terrain\s\w+', c.DisplayLabel);

REGEXP_EXTRACT(ECClassDef.[DisplayLabel],'(\w+)\s+(\w+)','\2,\1')
-----------------------------------------------------------------
Boundary,Terrain
Breakline,Terrain
Drape,Terrain
Drape,Terrain
Hole,Terrain
Island,Terrain
Reference,Terrain
Source,Terrain
Spot,Terrain
Void,Terrain
```

## StrToGuid( _guid-string_ )

> When `GUID` is stored a binary, it need to be converted for comparison purpose.

Reviewer: ```suggestion
When `GUID` is stored as a binary, it needs to be converted for comparison purposes.
```

```sql
SELECT * FROM [BisCore].[Element] WHERE FederationGuid = StrToGuid('407bfa18-944d-11ee-b9d1-0242ac120002')
```

## GuidToString( _binary-guid_ )

> When `GUID` is stored a binary, it need to be converted for comparison purpose.

Reviewer: ```suggestion
When `GUID` is stored as a binary, it needs to be converted for comparison purpose.
```
